libdrgn/kallsyms.c

```diff
@@ -0,0 +1,611 @@
+ * exists, then we use its first character to determine our behavior: 1, y, Y
+ * all indicate that we should use long names. 0, n, N all indicate that we
+ * should not.
+ */
+static bool guess_long_names(struct drgn_program *prog)
+{
+	const char *env = getenv("DRGN_KALLSYMS_LONG");
+	if (env) {
+		if (*env == '1' || *env == 'y' || *env == 'Y')
+			return true;
+		else if (*env == '0' || *env == 'n' || *env == 'N')
+			return false;
+	}
+
+	char *p = prog->vmcoreinfo.osrelease;
+	long major = strtol(p, &p, 10);
+	long minor = 0;
+	if (*p == '.')
+		minor = strtol(p + 1, NULL, 10);
+
+	return (major == 6 && minor >= 1) || major > 6;
+}
+
+/**
+ * Copy the kallsyms names tables from the program into host memory.
+ * @param prog Program to read from
+ * @param kr kallsyms_reader to populate
+ * @param vi vmcoreinfo for the program
+ */
+static struct drgn_error *
+kallsyms_copy_tables(struct drgn_program *prog, struct kallsyms_reader *kr,
+		     struct kallsyms_locations *loc)
+{
+	struct drgn_error *err;
+	const size_t token_index_size = (UINT8_MAX + 1) * sizeof(uint16_t);
+	uint64_t last_token;
+	size_t token_table_size, names_idx;
+	char data;
+	uint8_t len_u8;
+	int len;
+	bool bswap;
+
+	err = drgn_program_bswap(prog, &bswap);
+	if (err)
+		return err;
+
+	// Read num_syms from vmcore (bswap is done for us already)
+	err = drgn_program_read_u32(prog,
+				    loc->kallsyms_num_syms,
+				    false, &kr->num_syms);
+	if (err)
+		return err;
+
+	// Read the constant-sized token_index table (256 entries)
+	kr->token_index = malloc(token_index_size);
+	if (!kr->token_index)
+		return &drgn_enomem;
+	err = drgn_program_read_memory(prog, kr->token_index,
+				       loc->kallsyms_token_index,
+				       token_index_size, false);
+	if (err)
+		return err;
+	if (bswap)
+		for (size_t i = 0; i < kr->num_syms; i++)
+			kr->token_index[i] = bswap_16(kr->token_index[i]);
+
+	// Find the end of the last token, so we get the overall length of
+	// token_table. Then copy the token_table into host memory.
+	last_token = loc->kallsyms_token_table + kr->token_index[UINT8_MAX];
+	do {
+		err = drgn_program_read_memory(prog, &data,
+					       last_token, 1, false);
+		if (err)
+			return err;
+
+		last_token++;
+	} while (data);
+	token_table_size = last_token - loc->kallsyms_token_table + 1;
+	kr->token_table = malloc(token_table_size);
+	if (!kr->token_table)
+		return &drgn_enomem;
+	err = drgn_program_read_memory(prog, kr->token_table,
+				       loc->kallsyms_token_table,
+				       token_table_size, false);
+	if (err)
+		return err;
+
+	// Now find the end of the names array by skipping through it, then copy
+	// that into host memory.
+	names_idx = 0;
+	kr->long_names = guess_long_names(prog);
+	for (size_t i = 0; i < kr->num_syms; i++) {
+		err = drgn_program_read_u8(prog,
+					   loc->kallsyms_names + names_idx,
+					   false, &len_u8);
+		if (err)
+			return err;
+		len = len_u8;
+		if ((len & 0x80) && kr->long_names) {
```

Comment:
Considering that the max we could represent with 2 bytes of ULB128 is actually even lower (2^14 == 16k) and the bad track record for these sorts of limits, I definitely get your concern. Though I think there's an equally valid concern that, since kallsyms is a bespoke format, it may not make sense to support anything beyond what the existing kallsyms generator & consumer support. From the existing code, we have no real guarantee that they would use ULEB128 with three bytes, versus making some more substantial change to the format to accommodate larger symbol names.

I might go ahead and add a comment, but also raise an error if the MSB of the second byte is set (this would indicate a third byte is expected). That way we can be made aware of format breakages, but we don't make any assumptions about the future behavior of kallsyms.

Let me know if you disagree with the approach though.

Reply:
I didn't look at the kernel kallsyms code long enough to realize that they are also hard-coding at most 2 bytes despite describing it as ULEB128, so I'm fine with the comment either with or without the error.